changes/ticket21349

```diff
@@ -0,0 +1,6 @@
+  o Code simplification and refactoring:
```

Comment:
General comment: This branch needs to be rebased to the latest master, which contains the various modularization structure improvements. Let me know if you don't want to do this, and I will do it myself.

Reply:
I will do this once I address all the things you mentioned.